### ISSUE TITLE
Can't use Images in the basic tilemap tutorial.

### ISSUE THREAD

**Issue by minerscale** [kubilus1/gendev#27]
I've given up at this point... It works in SGDK but not Gendev.

I'm hoping someone can help me here...

So I have the bitmap tutorial with the moon. But I can't seem to get the damn thing to compile... :(

I'm getting this error.
`
src/main.c: In function ‘main’:
src/main.c:12:17: error: ‘Image {aka const struct <anonymous>}’ has no member named ‘w’
     u16 w = moon.w;
                 ^
src/main.c:14:17: error: ‘Image {aka const struct <anonymous>}’ has no member named ‘h’
     u16 h = moon.h;
                 ^
src/main.c:24:29: error: ‘Image {aka const struct <anonymous>}’ has no member named ‘image’
     VDP_loadBMPTileData(moon.image, 1, w / 8, h / 8, w / 8);
`

This also happens with the projects inside the sgdk-read-only sample directory.

Any help with me getting SGDK up and running on Linux would be greatly appreciated.

Cheers! :clinking_glasses:

**Comment by minerscale** [comment]
Ok... Spend a couple of hours working on a solution and literally 9 minuets after you file a bug report for it you find a solution... Is this because gendev uses a different version of SGDK? If so, I wonder where I could find a tutorial for this version?

By the way I substituted IMAGE for BITMAP.

**Comment by kubilus1** [comment]
Differences in SGDK version may be your culprit.  You should be able to pull in other SGDK versions with the gendev build chain since the steps are completely separate.  I have not personally tested the very latest SGDK though.

Also, in the past some of the tutorials have gone a bit out of date.  I recall that this one did not function at one time, but it is possible it is updated by now.

**Comment by minerscale** [comment]
How would I go about doing this?

Download a copy of SGDK and put it into sgdk/sgdk-read-only?

I would love to have the latest and greatest version of SGDK for my game.

EDIT: Trying to edit the makefile for SGDK to make it work with the latest version now. Hope it works.

Also: Why does this take so long to compile? It never goes flat out with the CPU or anything...
EDIT: Nevermind...
![ok](https://user-images.githubusercontent.com/20499616/27378976-ed3143ce-56bc-11e7-9277-c281e65cddb3.png)

EDIT again: I did it! I'm now running SGDK 1.3 in Linux! Thanks for pointing me in the right direction with changing versions.

**Comment by minerscale** [comment]
I 'spose It makes sense to close this now. welp. closing.

Shit... This didn't go as smoothly as I first though it did...

Now I can't get sprites working.
This along a whole slew of errors I ended up getting at compile time...
`main.c:(.text+0x3a): undefined reference to 'SPR_addSprite'`

Any ideas? thanks for the help.

**Comment by minerscale** [comment]
I suppose I'm asking, how do I port gendev to SGDK 1.3 and I also guess that that would take a bit of effort...

**Comment by kubilus1** [comment]
Yeah, you went the correct route with changing the sgdk/Makefile to clone https://github.com/Stephane-D/SGDK.git.  I would have to look at the the particular examples to see why they didn't work. It would not surprise me in the least to find out there are significant differences in the new SGDK leading to issues.

**Comment by kubilus1** [comment]
I suspect that part of the issue with SGDK 1.3 is rescomp, but there could be other issues as well.

**Comment by minerscale** [comment]
huh interesting... I would be happy to help figuring out a new version if that's something you would like me to do.

edit: downloaded the latest SGDK and rebuilt rescomp using code::blocks I hope this works out! While I'm at it, I may as well rebuild all the tools hey

edit 2: O_O an error while compiling appack! A syntax error... Oh my... windows specific code! Calling conventions! I'll give just removing the __cdecl from it and hope for the best... ooh! commented out code! mabye that will solve my woes! Nope... Fuck it

Damn it, I don't know a thing about this.

**Comment by kubilus1** [comment]
Yeah, appack was trouble.  As you can see in my project, I have patches against a handful of tool SGDK relies on.  Several are not very cross platform friendly, unfortunately.

Going code::blocks to Make isn't that bad since there is a cbp2make conversion tool.

Patches are always welcome if you get something going.

**Comment by minerscale** [comment]
So let me get this straight. There is nothing wrong with the GCC toolchain for SGDK 1.3, it's only a select few of the tools that won't properly build for linux right?

So what we have to do is find a way to port all the latest versions of the tools to Linux and we're golden.

It's just Stef didn't make the tools particularly portable.

Edit: I got the latest version of rescomp working on Linux! I needed to include lz4w.jar in the same directory though. Or it would segfault. But other than that. It's exactly the same as the original rescomp supplied by Stef. So there's a start. Do you know of any way to put lz4w.jar in a bin directory and only refer to it once? That seems to be what your doing. I also got sizebnd working perfectly with no changes to the code. Seems like wavtoraw works perfectly.

I tried XGM tool and tried the optimize part. But the out VGM was ~300 bytes _bigger_ than the input! In all honesty though it worked perfectly.

Yes baby! I got every tool to compile correctly after some dicking about. I now have a bin directory full of the new executables that I've tested. I hope I can get SGDK to work after that!

Hey, looking at the files directory I can see that you have a bunch of tarballs stored there. I can also see that they are 2 years old. Has that caused any problems with using stale versions of libraries or has it helped? I don't think this is the issue though.

I believe it would be a good idea if we just build the binaries and copy them over instead of faffing around with .diff files and all that. It reduces chances of error and if we specify a version of SGDK it doesn't matter anyway.

Also, everytime I want to test build gendev I have to compile GCC again. This takes ~1 hour on my computer. How do I not do that?

Alright. After all this. It's still not working... I don't know what's happened but SGDK refuses to work. It's the same error as before. I think the problem is a little more sinister than the tools. You see, I can't get any files to include into the project, except genesis.h So perhaps the problem is with a missing -I or something? I can't tell...

**Comment by kubilus1** [comment]
So this project has a couple parts to it.  There is the Genesis m68k GCC tool chain as the main component.  This is what allows compilation of stuff onto the Sega Genesis.  Projects can just use this toolchain, or libraries could be built up from this point.

SGDK is a separate thing.  Stef's project is developed specifically to run on Windows.  For the most part the code isn't that difficult to port, but some of the tools that are used may or may not be less friendly.  I have included a number of patches to the SGDK project in order to get it to work with the Linux GCC toolchain. 

The Linux GCC tool chain and the SGDK libs can be built separately, there is really no need to rebuild 100% of everything every time.
```
$ cd gendev
$ make sgdk_clean sgdk_build
```
 ... for instance


As far as 1.3 SGDK goes, it doesn't look like rescomp, or appack have changed, AFAICT.  There are some differences, of course with the SGDK libs themselves, and some of the supporting build components (Likely one of those build or linker changes is part of the issue you are seeing.)  My hope is that I can better minimize *my* patches to SGDK in order to reduce the difficulty in following that project.

On top of that, I figure it's kind of unnecessary to require rebuilding the GCC toolchain from an end user perspective, for the most part.  That is why I have uploaded the occasional release of the build chain.  Also, that is why I plugged this project up to Docker Hub, so that a docker container of the chain can be easily used.

At the moment, I've been working hard on getting the toolchain to build under Travis-CI.  This should hopefully automate the regular creation of releases of the toolchain, as well as validating patches.

As far as diff files, patches, etc.  That's kind of (half) the point of this project.  Figure out the steps in a repeatable way, capture and share the results.
